### PR TITLE
Type primitive arrays in BinaryFormattedObject

### PR DESCRIPTION
**Pull request by JeremyKuhne**
`ArraySinglePrimitive` was taking `IReadOnlyList<object>` which necessitated boxing all items. This converts `ArrayRecord` to be typed so we can avoid this and adds special optimizations for `byte[]`. This is important for performance of the types we're supporting or intending to support (such as `ImageListStreamer`).

It may be possible to further improve this. If the array data is always aligned in the input then we can just read/write the data directly as byte arrays. `DateTime`, `TimeSpan`, and `decimal` need individually parsed.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/10738)

### CODE REVIEW

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/BinaryFormat/Record.cs [hunk 1]
```diff
@@ -120,10 +198,89 @@ private protected static void WritePrimitiveTypes(BinaryWriter writer, Primitive
...
+        for (int i = 0; i < values.Count; i++)
+        {
+            if (typeof(T) == typeof(bool))
+            {
+                writer.Write((bool)(object)values[i]);
+            }
+            else if (typeof(T) == typeof(byte))
+            {
+                writer.Write((byte)(object)values[i]);
+            }
```
Comment (lonitra) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/BinaryFormat/Record.cs:220]: Do we still need this case now that we have the special casing above?

**Review by lonitra — APPROVED** [review]
👍

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/BinaryFormat/ArraySinglePrimitive.cs [hunk 2]
```diff
@@ -13,30 +13,28 @@ namespace System.Windows.Forms.BinaryFormat;
 ///   </see>
 ///  </para>
 /// </remarks>
-internal sealed class ArraySinglePrimitive : ArrayRecord, IRecord<ArraySinglePrimitive>
+internal sealed class ArraySinglePrimitive<T> : ArrayRecord<T>, IRecord<ArraySinglePrimitive<T>>, IPrimitiveTypeRecord
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/BinaryFormat/ArraySinglePrimitive.cs:16]: I do not want us naming any files as `~` anything.

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/BinaryFormat/Support/TypeInfo.cs [hunk 3]
```diff
@@ -84,4 +84,28 @@ internal static class TypeInfo
         // TypeCode.DBNull => 0,
         _ => type == typeof(TimeSpan) ? PrimitiveType.TimeSpan : default,
     };
+
+    /// <summary>
+    ///  Returns the <see cref="Type"/> for the given <paramref name="primitiveType"/>.
+    /// </summary>
+    internal static Type GetPrimitiveTypeType(this PrimitiveType primitiveType) => primitiveType switch
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/BinaryFormat/Support/TypeInfo.cs:91]: Deliberately did this to not collide with `object.GetType()`.

**Pull request merged**